File: Data Manipulation.py
### NOTES #####################################################################

#Uses environment 'Reserving-RNN-PyTorch'
#python 3.10.13
#pytorch 2.2.1
#torchvision 0.17.1
#cuda 12.1.0
#pandas 2.1.1
#numpy 1.26.0

# Takes simulated SPLICE output from R, converts it into a 'set' and 'index' 
# csv file, splits into training and test sets and stores them in csvs

### PACKAGES ##################################################################

import pandas as pd
import numpy as np
from random import sample, seed
import pathlib

### FUNCTIONS #################################################################

def manipulate_and_split(fp_in, fp_out, fp_out_noInc):
    '''Main function for manipulating the raw R output and splitting it into 
    training, validation and test sets ready for the RNN'''

    ### Helper functions ###
    def get_ultimate(row):
        '''Finds the ultimate claim cost for the given claim for a particular row'''
        return data.loc[data['claim_no'] == row['claim_no'], ['cumpaid']].iloc[-1]

    def get_incurred(row):
        '''Finds the estimated incurred claim cost for the given claim for a 
        particular row'''
        return data.loc[(data['claim_no'] == row['claim_no']) & 
                        (data['txn_time'] < row['c']), ['incurred']].iloc[-1]

    def get_cumpaid(row):
        return payment_rows.loc[(payment_rows['claim_no'] == row['claim_no']) & 
                                (payment_rows['txn_time'] <= row['c']), ['cumpaid']].max()

    def get_case_estimate(row):
        temp = data.loc[(data['claim_no'] == row['claim_no']) & 
                        (data['txn_time'] <= row['c'])]
        
        return temp.loc[(temp['txn_time'] == max(temp['txn_time'])), 
                        'incurred'].iloc[0]

    def get_finalised_quarter(row):
        return data.loc[data['claim_no'] == row['claim_no'], ['txn_time']].iloc[-1]

    def get_legal_rep(row):
        return data.loc[(data['claim_no'] == row['claim_no']) & 
                        (data['txn_time'] < row['c']), ['Legal Representation']].iloc[-1]

    def get_injury_sev(row):
        return data.loc[(data['claim_no'] == row['claim_no']) & 
                        (data['txn_time'] < row['c']), ['Injury Severity']].iloc[-1]
        
    def get_age(row):
        return data.loc[(data['claim_no'] == row['claim_no']) & 
                        (data['txn_time'] < row['c']), ['Age of Claimant']].iloc[-1]

    def get_acc_quarter(row):
        return data.loc[data['claim_no'] == row['claim_no'], ['acc_quarter']].iloc[0]


    ### Main function ###


    # Reading data
    data = pd.read_csv(fp_in)

    #data.head()

    # Dropping first column
    data = data.iloc[:, 1:]

    # Adding duration and start time columns, and rounding them
    start_group = data.groupby('claim_no')['txn_time'].min()
    start_group = np.ceil(start_group)

    end_group = data.groupby('claim_no')['txn_time'].max()
    end_group = np.ceil(end_group)

    # Adding raw payments to old dataframe
    payments = data['cumpaid'].diff().fillna(0)
    payments[payments < 0] = 0
    data['payments'] = payments

    # Adding finalisation time to old dataframe
    data['finalised_quarter'] = np.ceil(data.apply(get_finalised_quarter, axis = 1))

    # Adding censoring times to index dataframe
    index_data_list = []

    for claim_no in range(1, len(start_group) + 1):
        for c in range(int(start_group[claim_no]), int(end_group[claim_no])):
            dev_quarter = c - start_group[claim_no] + 1
            rept_quarter = start_group[claim_no]
            finalised_quarter = data.loc[data['claim_no'] == claim_no, 'finalised_quarter'].mean()
            index_data_list.append([claim_no, c, dev_quarter, rept_quarter, finalised_quarter])
            
    index_data = pd.DataFrame(index_data_list, columns = ['claim_no', 'c', 
                                                        'dev_quarter', 
                                                        'rept_quarter', 
                                                        'finalised_quarter'])       

    # Extracting claim size, incurred and current case estimate values
    index_data['claim_size'] = index_data.apply(get_ultimate, axis = 1)
    index_data['incurred'] = index_data.apply(get_incurred, axis = 1)
    index_data['case_estimate'] = index_data.apply(get_case_estimate, axis = 1)

    index_data['Legal Representation'] = index_data.apply(get_legal_rep, axis = 1)
    index_data['Injury Severity'] = index_data.apply(get_injury_sev, axis = 1)
    index_data['Age of Claimant'] = index_data.apply(get_age, axis = 1)

    index_data['acc_quarter'] = index_data.apply(get_acc_quarter, axis = 1)

    # Finding rows in original dataset that involve payments
    payment_rows = data.loc[(data['txn_type'] == 'P') | 
                            (data['txn_type'] == 'PMi') | 
                            (data['txn_type'] == 'PMa')]

    # Adding payment data
    index_data['cumpaid'] = index_data.apply(get_cumpaid, axis = 1)

    # Adding indexes to index set
    index_data['index'] = index_data.index

    # Adding m(t) and log(m(t)) to index dataset
    index_data['m'] = index_data['claim_size'] / index_data['case_estimate']
    index_data['log_m'] = np.log(index_data['m'])

    # Creating dataframe with only censored rows
    databoxes_list = []

    # Loop over indexes
    for i in range(len(index_data)):

        # subset claims from original dataset
        temp = data.loc[data['claim_no'] == index_data.iloc[i]['claim_no'], 
                        ['claim_no', 'txn_time', 'txn_delay', 'txn_type', 
                        'incurred', 'OCL', 'cumpaid']]

        # Loop over rows in original dataset
        for j in range(len(temp)):
            if (index_data.iloc[i]['claim_no'] == temp.iloc[j]['claim_no'] and 
                index_data.iloc[i]['c'] > temp.iloc[j]['txn_time']):

                databoxes_list.append([i] + 
                                    [index_data.iloc[i]['c']] + 
                                    list(temp.iloc[j]))

    # converting list to dataframe
    databoxes = pd.DataFrame(databoxes_list, columns = ['index', 'c', 'claim_no', 
                                                        'txn_time', 'txn_delay', 
                                                        'txn_type', 'incurred', 
                                                        'OCL', 'cumpaid'])

    index_data['cumpaid'] = index_data['cumpaid'].fillna(0)

    # adding log claim size to index data
    index_data['log_claim_size'] = np.log(index_data['claim_size'])

    # adding true_ocl to index data
    index_data['true_ocl'] = index_data['claim_size'] - index_data['cumpaid']
    index_data['log_true_ocl'] = np.log(index_data['true_ocl'])

    # renaming columns
    index_data.rename(columns = {'c': 'pred_time'}, inplace = True)
    databoxes.rename(columns = {'c': 'pred_time', 'txn_time': 'cal_time', 
                                'txn_delay': 'dev_time'}, inplace = True)

    # Replacing NAs with -1
    index_data = index_data.fillna(-1)
    databoxes = databoxes.fillna(-1)

    # rename columns
    databoxes.rename(columns = {'cumpaid': 'paid', 'OCL': 'ocl'}, inplace = True)
    index_data.rename(columns = {'incurred': 'latest_incurred'}, inplace = True)

    # Some claim reports have payments really close to 0, need to set these to exactly 0
    databoxes.loc[databoxes['paid'] < 0.01, 'paid'] = 0

    # Binary encode legal rep covariate
    index_data['Legal Representation'].replace({'Y': 1, 'N': 0}, inplace=True)

    # Ordinal encode age covariate
    index_data['Age of Claimant'].replace({'0-15': 0, '15-30': 2, '30-50': 2, '50-65': 3, 'over 65': 4}, inplace=True)

    # Make injury severity go from 0-5 instead of 1-6 so it works with embeddings
    index_data['Injury Severity'] = index_data['Injury Severity'] - 1

    ### TRAIN TEST SPLIT ##########################################################
    val_start_quarter = 36
    test_start_quarter = 40

    # Valuation date is 40
    train_index = index_data.loc[index_data['finalised_quarter'] <= val_start_quarter]
    val_index = index_data.loc[(index_data['finalised_quarter'] > val_start_quarter) & (index_data['finalised_quarter'] <= test_start_quarter)]
    test_index = index_data.loc[index_data['finalised_quarter'] > test_start_quarter]

    # TESTING: removing all observations in test sets that occur before the final observation in the training set
    val_index = val_index.loc[val_index['pred_time'] >= val_start_quarter]
    test_index = test_index.loc[test_index['pred_time'] >= test_start_quarter]


    train_set = databoxes.loc[databoxes['index'].isin(train_index['index'])]
    val_set = databoxes.loc[databoxes['index'].isin(val_index['index'])]
    test_set = databoxes.loc[databoxes['index'].isin(test_index['index'])]

    # creating noIncurred versions
    train_set_noInc = train_set.loc[((train_set['txn_type'] != 'Ma') | (train_set['paid'] == 0)) & (train_set['txn_type'] != 'Mi')]
    val_set_noInc = val_set.loc[((val_set['txn_type'] != 'Ma') | (val_set['paid'] == 0)) & (val_set['txn_type'] != 'Mi')]
    test_set_noInc = test_set.loc[((test_set['txn_type'] != 'Ma') | (test_set['paid'] == 0)) & (test_set['txn_type'] != 'Mi')]

    # creating different sets for different model input versions

    train_set = train_set.loc[:, ['index', 'claim_no', 'pred_time', 'dev_time', 
                                    'cal_time', 'paid', 'ocl']]

    val_set = val_set.loc[:, ['index', 'claim_no', 'pred_time', 'dev_time', 
                                'cal_time', 'paid', 'ocl']]

    test_set = test_set.loc[:, ['index', 'claim_no', 'pred_time', 'dev_time', 
                                'cal_time', 'paid', 'ocl']]

    train_index = train_index.loc[:, ['index', 'claim_no', 'pred_time', 
                                        'dev_quarter', 'rept_quarter', 'acc_quarter',
                                        'finalised_quarter', 'claim_size', 'log_claim_size',
                                        'latest_incurred', 'm', 'log_m',
                                        'true_ocl', 'log_true_ocl',
                                        'Legal Representation', 'Injury Severity',
                                        'Age of Claimant']]

    val_index = val_index.loc[:, ['index', 'claim_no', 'pred_time', 
                                    'dev_quarter', 'rept_quarter', 'acc_quarter',
                                    'finalised_quarter', 'claim_size', 'log_claim_size',
                                    'latest_incurred', 'm', 'log_m',
                                    'true_ocl', 'log_true_ocl',
                                    'Legal Representation', 'Injury Severity',
                                    'Age of Claimant']]

    test_index = test_index.loc[:, ['index', 'claim_no', 'pred_time', 
                                    'dev_quarter', 'rept_quarter', 'acc_quarter',
                                    'finalised_quarter', 'claim_size', 'log_claim_size',
                                    'latest_incurred', 'm', 'log_m',
                                    'true_ocl', 'log_true_ocl',
                                    'Legal Representation', 'Injury Severity',
                                    'Age of Claimant']]

    train_set_noInc = train_set_noInc.loc[:, ['index', 'claim_no', 'pred_time', 
                                                'dev_time', 'cal_time', 'paid', 
                                                'ocl']]

    val_set_noInc = val_set_noInc.loc[:, ['index', 'claim_no', 'pred_time', 
                                            'dev_time', 'cal_time', 'paid', 'ocl']]

    test_set_noInc = test_set_noInc.loc[:, ['index', 'claim_no', 'pred_time', 
                                        'dev_time', 'cal_time', 'paid', 'ocl']]


    # Exporting With Incurred
    train_index.to_csv(fp_out + 'train_index.csv', index = False)
    val_index.to_csv(fp_out + 'val_index.csv', index = False)
    test_index.to_csv(fp_out + 'test_index.csv', index = False)

    train_set.to_csv(fp_out + 'train_set.csv', index = False)
    val_set.to_csv(fp_out + 'val_set.csv', index = False)
    test_set.to_csv(fp_out + 'test_set.csv', index = False)

    # Exporting No Incurred
    train_index.to_csv(fp_out_noInc + 'train_index.csv', index = False)
    val_index.to_csv(fp_out_noInc + 'val_index.csv', index = False)
    test_index.to_csv(fp_out_noInc + 'test_index.csv', index = False)

    train_set_noInc.to_csv(fp_out_noInc + 'train_set.csv', index = False)
    val_set_noInc.to_csv(fp_out_noInc + 'val_set.csv', index = False)
    test_set_noInc.to_csv(fp_out_noInc + 'test_set.csv', index = False)


### FILEPATHS AND FUNCTION CALLS (ADJUST THESE BEFORE RUNNING) ################

# creating 1 dataset
#fp_in = 'Datasets/R Outputs/data_noInf_cov_TRUE_seed_20250101.csv'
#fp_out = 'Datasets/Python Inputs/noInf_WithInc_seed_20250101/'
#fp_out_noInc = 'Datasets/Python Inputs/noInf_NoInc_seed_20250101/'

fp_in = 'Datasets/R Outputs/data_noInf_cov_TRUE_seed_20201006.csv'
fp_out = 'Datasets/Python Inputs/noInf_WithInc_seed_20201006/'
fp_out_noInc = 'Datasets/Python Inputs/noInf_NoInc_seed_20201006/'

print('file running...')
manipulate_and_split(fp_in, fp_out, fp_out_noInc)
print('file complete')

'''
# creating multiple datasets
max_iter = 10
seed_base = 2034

fp_R = 'Datasets/R Outputs/data_noInf_cov_TRUE_seed_'
fp_py_WithInc = 'Datasets/Python Inputs/noInf_WithInc_seed_'
fp_py_noInc = 'Datasets/Python Inputs/noInf_NoInc_seed_'

for i in range(1, max_iter + 1):
    fp_in = fp_R + str(i + seed_base) + '.csv'
    fp_out = fp_py_WithInc + str(i + seed_base) + '/'
    fp_out_noInc = fp_py_noInc + str(i + seed_base) + '/'
    
    print('Seed: ' + str(i + seed_base))
    #print('R fp: ' + fp_R_full)
    #print('Python fp: ' + fp_py_full)
    pathlib.Path(fp_out).mkdir(parents = True, exist_ok = True)
    pathlib.Path(fp_out_noInc).mkdir(parents = True, exist_ok = True)

    manipulate_and_split(fp_in, fp_out, fp_out_noInc)

print('DATA EXPORTED')
'''

File: Data Manipulation.R
# Takes simulated SPLICE output from R, converts it into a 'set' and 'index' 
# csv file, splits into training and test sets and stores them in csvs

library(data.table)
library(rstudioapi)

# Getting the path of current open file
current_path = rstudioapi::getActiveDocumentContext()$path 
setwd(dirname(current_path ))
print( getwd() )


################################################################################

data_manipulation <- function(fp_in, fp_out, fp_out_noInc) {

  # Reading data
  data <- fread(fp_in)
  
  # Dropping first column
  data[, V1 := NULL]
  
  
  # Adding duration and start time columns, and rounding them
  data[, ':=' (rept_quarter = ceiling(min(txn_time)),
               finalised_quarter = ceiling(max(txn_time))), keyby=claim_no]
  
  # remove claims that are reported and settled within the same quarter
  data <- data[!(claim_no %in% data[rept_quarter == finalised_quarter, unique(claim_no)])]
  
  # Adding incremental payments to data
  payments <- data[, diff(cumpaid)]
  payments[payments < 0] <- 0
  payments <- c(0, payments)
  
  data[, payment := payments]
  
  
  # Adding incremental case estimate revisions to data
  revisions <- data[, diff(incurred)]
  revisions <- c(0, revisions)
  
  data[, revision := revisions]
  data[txn_delay == 0, revision := incurred]
  
  data[revision == 0, .N]
  data[revision > 0 & txn_delay != 0, .N]
  
  
  # Creating index data
  index_data <- data[, .(pred_time = seq(mean(rept_quarter), mean(finalised_quarter) - 1)), keyby = claim_no]
  
  index_data <- merge(index_data, data[, .(acc_quarter = mean(acc_quarter), 
                                           rept_quarter = mean(rept_quarter), 
                                           finalised_quarter = mean(finalised_quarter), 
                                           claim_size = mean(claim_size), 
                                           `Legal Representation` = unique(`Legal Representation`), 
                                           `Injury Severity` = unique(`Injury Severity`), 
                                           `Age of Claimant` = unique(`Age of Claimant`)), keyby=claim_no], by="claim_no")
  
  # when using nested data table in condition, need column names to be different
  setnames(data, "claim_no", "claim_number")
  
  index_data[, ':=' (incurred = data[claim_number == claim_no & txn_time < pred_time, incurred[length(incurred)]],
                     cumpaid = data[claim_number == claim_no & txn_time < pred_time, cumpaid[length(cumpaid)]]), keyby=.(claim_no, pred_time)]
  
  # having $1 discrepancy is ok due to rounding 
  #View(index_data[incurred != OCL + cumpaid, .(claim_no, incurred, OCL, cumpaid, OCL + cumpaid)])
  
  index_data[, ':=' (index = 0:(.N - 1),
                     dev_quarter = pred_time - rept_quarter + 1,
                     m = claim_size / incurred,
                     log_m = log(claim_size / incurred),
                     true_ocl = claim_size - cumpaid,
                     log_claim_size = log(claim_size),
                     log_true_ocl = log(claim_size - cumpaid))]
  
  
  # Creating databoxes (i.e. 'set' data)
  databoxes <- index_data[, data[claim_number == claim_no & txn_time < pred_time, .(txn_time, txn_delay, cumpaid, OCL, txn_type, payment, revision)], keyby = .(index, claim_no, pred_time)]
  
  
  # renaming columns
  setnames(data, "claim_number", "claim_no")
  setnames(index_data, c("incurred"), c("latest_incurred"))
  setnames(databoxes, c('txn_time', 'txn_delay', 'cumpaid', 'OCL'), c('cal_time', 'dev_time', 'paid', 'ocl'))
  
  # Some claim reports have payments really close to 0, need to set these to exactly 0
  databoxes[paid > 0 & paid < 0.01, paid := 0]
  
  # Binary encode legal rep covariate
  index_data[`Legal Representation` == 'Y', `Legal Representation` := 1]
  index_data[`Legal Representation` == 'N', `Legal Representation` := 0]
  index_data[, `Legal Representation` := as.numeric(`Legal Representation`)]
  
  # Ordinal encode age covariate
  index_data[`Age of Claimant` == '0-15', `Age of Claimant` := 0]
  index_data[`Age of Claimant` == '15-30', `Age of Claimant` := 1]
  index_data[`Age of Claimant` == '30-50', `Age of Claimant` := 2]
  index_data[`Age of Claimant` == '50-65', `Age of Claimant` := 3]
  index_data[`Age of Claimant` == 'over 65', `Age of Claimant` := 4]
  index_data[, `Age of Claimant` := as.numeric(`Age of Claimant`)]
  
  # Make injury severity go from 0-5 instead of 1-6 so it works with embeddings
  index_data[, `Injury Severity` := `Injury Severity` - 1]
  
  # reordering index data columns
  setcolorder(index_data, c('index', 
                            'claim_no', 
                            'pred_time', 
                            'dev_quarter', 
                            'rept_quarter', 
                            'acc_quarter', 
                            'finalised_quarter', 
                            'claim_size', 
                            'log_claim_size', 
                            'latest_incurred', 
                            'm', 
                            'log_m', 
                            'true_ocl', 
                            'log_true_ocl', 
                            'Legal Representation', 
                            'Injury Severity', 
                            'Age of Claimant'))
  
  index_data[, cumpaid := NULL] # can add back later
  
  
  # Adding FNN summary data to index
  payment_summary <- databoxes[payment > 0, .(num_payments = .N,
                               mean_payments = mean(payment),
                               vco_payments = sd(payment) / mean(payment),
                               max_payment = max(payment)), keyby=index]
  
  index_data <- merge(index_data, payment_summary, by='index', all.x=TRUE)
  
  revision_summary <- databoxes[revision != 0 & dev_time > 0, .(num_revisions = .N,
                                               mean_revisions = mean(revision),
                                               max_revision = max(revision),
                                               total_revisions = sum(abs(revision)),
                                               prop_upward_revisions = sum(revision > 0) / .N), keyby=index]
  
  index_data <- merge(index_data, revision_summary, by='index', all.x=TRUE)
  
  index_data[is.na(index_data)] <- 0
  
  ### TRAIN TEST SPLIT ######################################################
  val_start_quarter = 36
  test_start_quarter = 40
  
  # Valuation date is 40
  train_index <- index_data[finalised_quarter <= val_start_quarter]
  val_index <- index_data[finalised_quarter > val_start_quarter & finalised_quarter <= test_start_quarter]
  test_index <- index_data[finalised_quarter > test_start_quarter]
  
  # TESTING: removing all observations in test sets that occur before the final observation in the training set
  val_index <- val_index[pred_time >= val_start_quarter]
  test_index <- test_index[pred_time >= test_start_quarter]
  
  
  # TESTING: randomly splitting by claim (i.e. not by time)
  # train_prop = 0.6
  # val_prop = 0.2
  # test_prop = 0.2
  # 
  # all_claims <- 1:max(index_data$claim_no)
  # 
  # train_claims <- sample(all_claims, train_prop * length(all_claims), replace = FALSE)
  # val_claims <- all_claims[!(all_claims %in% train_claims)]
  # 
  # test_claims <- sample(val_claims, test_prop / (val_prop + test_prop) * length(val_claims), replace = FALSE)
  # val_claims <- val_claims[!(val_claims %in% test_claims)]
  # 
  # train_index <- index_data[claim_no %in% train_claims]
  # val_index <- index_data[claim_no %in% val_claims]
  # test_index <- index_data[claim_no %in% test_claims]
  
  
  
  train_set <- databoxes[index %in% train_index[, index]]
  val_set <- databoxes[index %in% val_index[, index]]
  test_set <- databoxes[index %in% test_index[, index]]
  
  # creating noIncurred versions
  train_set_noInc = train_set[((txn_type != 'Ma') | (paid == 0)) & (txn_type != 'Mi')]
  val_set_noInc = val_set[((txn_type != 'Ma') | (paid == 0)) & (txn_type != 'Mi')]
  test_set_noInc = test_set[((txn_type != 'Ma') | (paid == 0)) & (txn_type != 'Mi')]
  
  # removing unnecessary columns
  train_set[, txn_type := NULL]
  val_set[, txn_type := NULL]
  test_set[, txn_type := NULL]
  
  train_set_noInc[, ':=' (txn_type = NULL, revision = NULL)]
  val_set_noInc[, ':=' (txn_type = NULL, revision = NULL)]
  test_set_noInc[, ':=' (txn_type = NULL, revision = NULL)]
  
  train_index_noInc <- train_index[, .SD, .SDcols = !c('num_revisions', 'mean_revisions', 'max_revision', 'total_revisions', 'prop_upward_revisions')]
  val_index_noInc <- val_index[, .SD, .SDcols = !c('num_revisions', 'mean_revisions', 'max_revision', 'total_revisions', 'prop_upward_revisions')]
  test_index_noInc <- test_index[, .SD, .SDcols = !c('num_revisions', 'mean_revisions', 'max_revision', 'total_revisions', 'prop_upward_revisions')]
  
  
  # Exporting With Incurred (check that not export with row column)
  fwrite(train_index, paste0(fp_out, 'train_index.csv'))
  fwrite(val_index, paste0(fp_out, 'val_index.csv'))
  fwrite(test_index, paste0(fp_out, 'test_index.csv'))
  
  fwrite(train_set, paste0(fp_out, 'train_set.csv'))
  fwrite(val_set, paste0(fp_out, 'val_set.csv'))
  fwrite(test_set, paste0(fp_out, 'test_set.csv'))
  
  # Exporting No Incurred
  fwrite(train_index_noInc, paste0(fp_out_noInc, 'train_index.csv'))
  fwrite(val_index_noInc, paste0(fp_out_noInc, 'val_index.csv'))
  fwrite(test_index_noInc, paste0(fp_out_noInc, 'test_index.csv'))
  
  fwrite(train_set_noInc, paste0(fp_out_noInc, 'train_set.csv'))
  fwrite(val_set_noInc, paste0(fp_out_noInc, 'val_set.csv'))
  fwrite(test_set_noInc, paste0(fp_out_noInc, 'test_set.csv'))
}

################################################################################

# fp_in = './Datasets/R Outputs/small sample.csv'
# fp_out = './Datasets/Python Inputs/small sample WithInc/'
# fp_out_noInc = './Datasets/Python Inputs/small sample NoInc/'


# fp_in = './Datasets/R Outputs/complexity 1.csv'
# fp_out = './Datasets/Python Inputs/complexity 1 WithInc/'
# fp_out_noInc = './Datasets/Python Inputs/complexity 1 NoInc/'

fp_in = './Datasets/R Outputs/data_noInf_cov_TRUE_seed_20201006.csv'
fp_out = './Datasets/Python Inputs/noInf_WithInc_seed_20201006/'
fp_out_noInc = './Datasets/Python Inputs/noInf_NoInc_seed_20201006/'


data_manipulation(fp_in, fp_out, fp_out_noInc)


# creating multiple datasets
max_iter = 20
seed_base = 2024

fp_R = './Datasets/R Outputs/data_noInf_cov_TRUE_seed_'
fp_py_WithInc = './Datasets/Python Inputs/noInf_WithInc_seed_'
fp_py_noInc = './Datasets/Python Inputs/noInf_NoInc_seed_'

for (i in 1:max_iter) {
  fp_in = paste0(fp_R, i + seed_base, '.csv')
  fp_out = paste0(fp_py_WithInc, i + seed_base, '/')
  fp_out_noInc = paste0(fp_py_noInc, i + seed_base, '/')

  print(paste0('Seed: ', i + seed_base))
  #print(paste0('fp_in: ', fp_in))
  #print(paste0('fp out: ', fp_out))

  data_manipulation(fp_in, fp_out, fp_out_noInc)
}
print("Done!")
